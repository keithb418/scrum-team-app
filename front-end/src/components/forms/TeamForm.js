import React from "react";
import { Form, FormGroup, ControlLabel, FormControl, Button } from "react-bootstrap";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import { withRouter } from "react-router";

import CancelButton from "./CancelButton";

class TeamForm extends React.Component {
  constructor (props, title, onSubmitAction) {
    super(props);
    this.state = {
      name: props.name || "",
      title: title || "Add a New Team"
    };
    this.onInputChange = this.onInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.onSubmitAction = onSubmitAction || this.props.createItem;
  }

  onInputChange (evt) {
    const target = evt.target;
    const value = target.value;
    const name = target.name;

    this.setState({
      [name]: value
    });
  };

  getValidationState () {
    const teamNameLength = this.state.name.length;
    if (teamNameLength > 0 && teamNameLength <= 50) {
      return "success";
    } else {
      return "error";
    }
  };

  handleSubmit () {
    const { name } = this.state;
    const validationState = this.getValidationState();

    if (validationState === "error") {
      return false;
    } else {
      if (this.state.title === "Edit Team") {
        let _id = this.props.teamId;
        this.onSubmitAction(_id, name);
        this.props.history.push("/");
      } else {
        this.onSubmitAction({ name });
        this.props.history.push("/");
      }
    }
  }

  render () {
    let buttonText = (this.state.title === "Edit Team") ? "Edit Team" : "Add Team";
    const isDisabled = this.getValidationState() === "error" ? this.state.isDisabled = true : this.state.isDisabled = false;

    return (
      <div className="row">
        <div className="form-section col-md-6 offset-md-3 col-xs-12">
          <div className="panel-heading">
            <h3 className="panel-title">{this.state.title}<sup><i className="fa fa-asterisk required"></i></sup></h3>
          </div>
          <hr />
          <Form>
            <FormGroup controlId="team">
              <FormControl
                type="text"
                name="name"
                value={this.state.name}
                placeholder="Enter team name"
                onChange={this.onInputChange}
              />
              <FormControl.Feedback />
            </FormGroup>
            <Button
              bsStyle="primary"
              disabled={isDisabled}
              onClick={this.handleSubmit}>
              {buttonText}
            </Button>
            <CancelButton />
          </Form>
        </div>
      </div>
    );
  };
};

TeamForm.propTypes = {
  handleSubmit: PropTypes.func
};

export default TeamForm;