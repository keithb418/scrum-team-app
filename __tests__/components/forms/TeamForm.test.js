import React from 'react';
import { shallow } from 'enzyme';
import TeamForm from '../../../front-end/src/components/forms/TeamForm';
import teamData from '../../fixtures/teams';

test('should render TeamForm correctly', () => {
    const wrapper = shallow(
        <TeamForm
            title="Test Team Form"
            team={teamData[0]}/>
    );
    expect(wrapper).toMatchSnapshot()
});

test('should render invalid TeamForm correctly', () => {
    const getValidationState = jest.fn().mockImplementation(() => "error");
    const wrapper = shallow(<TeamForm/>);
    wrapper.find('Form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('isDisabled')).toBe(true);
    expect(wrapper.find('Button').props().disabled).toBe(true);
    expect(wrapper).toMatchSnapshot();
});

test('should set name input change', () => {
    const wrapper = shallow(<TeamForm/>);
    const name = "name"
    const value = "new Team";
    wrapper.find('FormControl').simulate('change', {
        target: { name, value }
    });
    expect(wrapper.state('name')).toBe(value);
});

test('should call onSubmit prop for valid form submission', () => {
    const onSubmit = jest.fn();
    const getValidationState = jest.fn().mockImplementation(() => "success");
    const wrapper = shallow(
        <TeamForm
            title="Test Team Form"
            team={teamData[0]}
            onSubmit={onSubmit}/>
    );
    wrapper.find('Form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('isDisabled')).toBe(false);
    expect(wrapper.find('Button').props().disabled).toBe(false);
    expect(onSubmit).toHaveBeenLastCalledWith({
        name: teamData[0].name,
    });
});