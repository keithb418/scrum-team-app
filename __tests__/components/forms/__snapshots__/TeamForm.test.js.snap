// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TeamForm correctly 1`] = `
<div
  className="row"
>
  <div
    className="form-section col-md-6 offset-md-3 col-xs-12"
  >
    <div
      className="panel-heading"
    >
      <h3
        className="panel-title"
      >
        Test Team Form
        <sup>
          <i
            className="fa fa-asterisk required"
          />
        </sup>
      </h3>
    </div>
    <hr />
    <Form
      bsClass="form"
      componentClass="form"
      horizontal={false}
      inline={false}
      onSubmit={[Function]}
    >
      <FormGroup
        bsClass="form-group"
        controlId="team"
      >
        <FormControl
          bsClass="form-control"
          componentClass="input"
          name="name"
          onChange={[Function]}
          placeholder="Enter team name"
          type="text"
          value="Team 1"
        />
        <FormControlFeedback
          bsClass="form-control-feedback"
          bsRole="feedback"
        />
      </FormGroup>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        disabled={false}
        type="submit"
      >
        Test Team Form
      </Button>
      <CancelButton />
    </Form>
  </div>
</div>
`;

exports[`should render invalid TeamForm correctly 1`] = `
<div
  className="row"
>
  <div
    className="form-section col-md-6 offset-md-3 col-xs-12"
  >
    <div
      className="panel-heading"
    >
      <h3
        className="panel-title"
      >
        Add a New Team
        <sup>
          <i
            className="fa fa-asterisk required"
          />
        </sup>
      </h3>
    </div>
    <hr />
    <Form
      bsClass="form"
      componentClass="form"
      horizontal={false}
      inline={false}
      onSubmit={[Function]}
    >
      <FormGroup
        bsClass="form-group"
        controlId="team"
      >
        <FormControl
          bsClass="form-control"
          componentClass="input"
          name="name"
          onChange={[Function]}
          placeholder="Enter team name"
          type="text"
          value=""
        />
        <FormControlFeedback
          bsClass="form-control-feedback"
          bsRole="feedback"
        />
      </FormGroup>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        disabled={true}
        type="submit"
      />
      <CancelButton />
    </Form>
  </div>
</div>
`;
